### PR TITLE
AssetGroup.prefixed

### PR DESCRIPTION
**Pull request by sryza**
This proposes an approach to namespacing sets of assets, while ensuring that dependencies don't break.

### CODE REVIEW

File: python_modules/dagster/dagster/core/asset_defs/assets.py [hunk 1]
```diff
@@ -61,3 +77,28 @@ def get_partition_mapping(self, in_asset_key: AssetKey) -> PartitionMapping:
...
+
+    def with_replaced_asset_keys(
+        self,
+        output_asset_key_replacements: Mapping[AssetKey, AssetKey],
+        input_asset_key_replacements: Mapping[AssetKey, AssetKey],
+    ) -> "AssetsDefinition":
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=ExperimentalWarning)
+
+            return self.__class__(
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/asset_defs/assets.py:89]: these methods are always scary since its easy to introduce bugs where new optional fields on the class don't get added everywhere and then fail to get copied along. Having them in the same file like this is good. Maybe worth leaving comment notes under the last arg in the init signature to point towards these like [1]

File: python_modules/dagster/dagster/core/definitions/input.py [hunk 2]
```diff
@@ -273,6 +273,18 @@ def combine_with_inferred(
             asset_partitions=self._asset_partitions_fn,
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/definitions/input.py:273]: <i class='graphite__hidden'>[Re: line 86]<br/></i>

[1] update this (not sure how this renders on github directly, but its the last arg in InputDef init )<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7395?utm_source=unchanged-line-comment">Graphite</a>.</span>

**Review by alangenfeld — COMMENTED** [review]
how'd you arrive at these names? 

* feels like there should be a verb somewhere
* are we certain that asset_key prefixing is the only bulk update we want to do on a set of assets? If not would we want to chain methods or have a single entry with multiple arguments for updating?

**Review by alangenfeld — COMMENTED** [review]
just to clarify don't have that strong of a negative reaction of the names, but seemed worth discussing.

**Comment by sryza** [comment]
@alangenfeld 

> feels like there should be a verb somewhere

I basically chose the name for consistency with `configured`. I was worried that names like `prefix_assets` and `prefix` might suggest we're mutating the underlying objects instead of returning a new collection. That said, very open here if you have other suggestions. I'm particularly un-enamored with `assets_prefixed` right now because it just feels kind of hard to parse, so this PR keeps it out of the public API.

> are we certain that asset_key prefixing is the only bulk update we want to do on a set of assets? If not would we want to chain methods or have a single entry with multiple arguments for updating?

Yeah, good question. I definitely don't think we're certain of that. E.g. could imagine wanting to add tags/metadata as well.  I feel like chaining could look nice - though one potential downside is perf if you have lots of assets.

Do you have thoughts?

File: python_modules/dagster/dagster/core/definitions/output.py [hunk 3]
```diff
@@ -257,6 +257,19 @@ def combine_with_inferred(self: TOut, inferred: InferredOutputProps) -> TOut:
             asset_partitions_def=self.asset_partitions_def,
         )
 
+    def with_asset_key(self, asset_key: AssetKey) -> "OutputDefinition":
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/definitions/output.py:260]: luckily a lot of this painful stuff won't be necessary in the future (when we move away from OutputDef.asset_key

File: python_modules/dagster/dagster/core/asset_defs/assets.py [hunk 4]
```diff
@@ -37,6 +39,16 @@ def __call__(self, *args, **kwargs):
     def op(self) -> OpDefinition:
         return self._op
 
+    @property
+    def asset_key(self) -> AssetKey:
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/asset_defs/assets.py:43]: do we really need this method? it makes the docs a little cleaner, but are there other benefits?

Comment (sryza) [reply python_modules/dagster/dagster/core/asset_defs/assets.py:43]: I'm having a little trouble recalling, but I think there are a few situations where I've wished this was around. As a random example, imagine a user wants to write a utility that creates a job that does some sort of observing of a particular asset.

```
def make_asset_observer_job(the_asset):
    @job(name=f"observer_job_{'_'.join(the_asset.asset_key.path)}")
    def _observer_job():
        ...

    return _observer_job
```

We might argue that they should handle the fully general multi-asset case, but that might be overkill for them.


Imagine if a user wanted to write a utility that like created a job that

**Comment by OwenKephart** [comment]
I wonder if we should think about "AssetKeySuffix"s as a separate concept from an AssetKey, as I feel like using the same class for each of them might be causing more pain than it's worth. Ignoring the fact that OutputDefinitions require a literal AssetKey object (as this is temporary), I feel like it's at least a little bit confusing to create a bunch of data structures that reference AssetKeys, and not know if those are actual AssetKeys that will exist or if they're just the suffix of a larger asset key. I'd feel more comfortable if the python layer was actually able to distinguish between these cases (maybe even just using a tuple of strings for these unresolved asset keys, then maintaining a map between the tuple and the resolved key, where the map's the only thing that needs to update when you prefix a set of assets).

I also wonder about the ergonomics of having to manually specify the upstream groups. Do you have thoughts on the pros/cons of doing that vs. just waiting until we need to build a job out of the assets to resolve the input asset keys (i.e. waiting until we have the entire asset group, and using all of the resolved output asset keys as potential things to match with)? The tradeoff (to me) seems to be higher chance of key collision (because you're matching against all assets, not just some upstream subset) vs. easier interface (because you don't have to think about / track down the python object for the asset group that an asset you care about lives in)

**Comment by OwenKephart** [comment]
Previous comment got somewhat mangled, but overall I like the direction. Some thoughts:

I wonder if we should think about "AssetKeySuffix"s as a separate concept from an AssetKey, as I feel like using the same class for each of them might be causing more pain than it's worth. Ignoring the fact that OutputDefinitions require a literal AssetKey object (as this is temporary), I feel like it's at least a little bit confusing to create a bunch of data structures that reference AssetKeys, and not know if those are actual AssetKeys that will exist or if they're just the suffix of a larger asset key. I'd feel more comfortable if the python layer was actually able to distinguish between these cases (maybe even just using a tuple of strings for these unresolved asset keys, then maintaining a map between the tuple and the resolved key, where the map's the only thing that needs to update when you prefix a set of assets).

I also wonder about the ergonomics of having to manually specify the upstream groups. Do you have thoughts on the pros/cons of doing that vs. just waiting until we need to build a job out of the assets to resolve the input asset keys (i.e. waiting until we have the entire asset group, and using all of the resolved output asset keys as potential things to match with)? The tradeoff (to me) seems to be higher chance of key collision (because you're matching against all assets, not just some upstream subset) vs. easier interface (because you don't have to think about / track down the python object for the asset group that an asset you care about lives in)

File: python_modules/dagster/dagster/core/asset_defs/assets.py (hunk 4 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/core/asset_defs/assets.py:43]: I'm not against it, happy to have it stick around

File: python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_group.py [hunk 5]
```diff
@@ -521,3 +524,121 @@ def follows_o2(o2):
...
+    def asset1():
+        ...
+
+    upstream_group = AssetGroup([asset1]).prefixed("core")
+
+    @asset(name="apple")
+    def asset2():
+        ...
+
+    @asset(ins={"apple": AssetIn(namespace="core")})
```
Comment (clairelin135) [inline python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_group.py:581]: We probably will need to change `namespace` -> `prefix` in a later PR.

File: python_modules/dagster/dagster/core/asset_defs/asset_group.py [hunk 6]
```diff
@@ -508,6 +513,86 @@ def materialize(
             name="in_process_materialization_job", selection=selection
         ).execute_in_process()
 
+    def prefixed(self, key_prefix: str, upstream_groups: Optional[Sequence["AssetGroup"]] = None):
```
Comment (clairelin135) [inline python_modules/dagster/dagster/core/asset_defs/asset_group.py:516]: I think with this method it's possible to prefix an AssetGroup multiple times, e.g.:
```
group_1 = AssetGroup([asset1]).prefixed("prefix_1")
group_2 = group_1.prefixed("prefix_2")
```
and end up with an AssetKey like `AssetKey(['prefix_2', 'prefix_1', 'asset1'])`.
How will we know which asset to select for a certain prefix `AssetIn(prefix='prefix_1')`? It's unclear which asset this `AssetIn` would be referring to. Maybe `AssetIn` needs to accept a list of prefixes and match starting from the most top level prefix?

**Comment by sryza** [comment]
@OwenKephart 

> I wonder if we should think about "AssetKeySuffix"s as a separate concept from an AssetKey, as I feel like using the same class for each of them might be causing more pain than it's worth. Ignoring the fact that OutputDefinitions require a literal AssetKey object (as this is temporary), I feel like it's at least a little bit confusing to create a bunch of data structures that reference AssetKeys, and not know if those are actual AssetKeys that will exist or if they're just the suffix of a larger asset key. I'd feel more comfortable if the python layer was actually able to distinguish between these cases (maybe even just using a tuple of strings for these unresolved asset keys, then maintaining a map between the tuple and the resolved key, where the map's the only thing that needs to update when you prefix a set of assets).

Interesting. Do you have situations in mind where this pain is most likely to manifest?

I think this raises the question of when we decide that an asset key has been fully "resolved". Do you have thoughts? Some scattered thoughts from me:
* Once the asset key has made it inside a job, I think we can definitely argue that it's fully resolved. However, we want to allow working with assets without thinking about jobs.
* So what about when it's on an asset group? This gets hairy when you think about wanting to nest an asset group inside another asset group. And I think life becomes a lot simpler if you can think of an asset group as just a thin wrapper around a list of asset definitions, rather than introducing a new set of semantics.
* We could introduce a new concept of a `ResolvedAssetGroup` that's meant to be a terminal asset group. Feels kind of hairy though.

The way of looking at it that I find myself gravitating towards is that "fully resolving" an asset key basically happens at the point that it's placed inside a repository. But at that point, we're kind of beyond the Python APIs, so a distinction within the Python APIs might have limited value.

> I also wonder about the ergonomics of having to manually specify the upstream groups. Do you have thoughts on the pros/cons of doing that vs. just waiting until we need to build a job out of the assets to resolve the input asset keys (i.e. waiting until we have the entire asset group, and using all of the resolved output asset keys as potential things to match with)? The tradeoff (to me) seems to be higher chance of key collision (because you're matching against all assets, not just some upstream subset) vs. easier interface (because you don't have to think about / track down the python object for the asset group that an asset you care about lives in)

I think your assessment of the tradeoff is accurate. The main thing that drove me away from that approach is basically that avoiding collisions is one of the main aims of prefixing. If our solution doesn't address it pretty comprehensively, I think we're not meeting the goal. If keys don't get resolved until full asset group time, then, in a sufficiently large org, I think everyone basically ends up needing to prefix everything to defend against collisions.

FWIW, the approach here is basically ripped off from the way imports and namespacing work in most programming languages:
* When you're in a Python module, you don't need to worry about having names that are globally unique across all modules in your git repository.
* "upstream_groups" are the analog of imports - you use them to bring a set of definitions into scope, which allows linking against the right ones.

File: python_modules/dagster/dagster/core/asset_defs/asset_group.py (hunk 6 above)
Comment (sryza) [reply python_modules/dagster/dagster/core/asset_defs/asset_group.py:516]: Here's how it would work with the approach in this PR:

Let's say we've got an asset with an `AssetIn` with a prefix:
```
@asset(ins={"asset1": AssetIn(prefix="prefix1"))
def asset2(asset1):
    ...
```

That's exactly equivalent to the following:
```
@asset(ins={"asset1": AssetIn(asset_key=AssetKey(["prefix1", "asset1"])})
def asset2(asset1):
    ...
```
I.e. `asset2.dep_asset_keys == {AssetKey(["prefix1", "asset1"])}`

Now let's say we have an `AssetGroup` called `group1`  with an asset with `AssetKey(['prefix_2', 'prefix_1', 'asset1'])`:
```
assert group1.assets[0].asset_key == AssetKey(['prefix_2', 'prefix_1', 'asset1'])
```

Now we want to put `asset2` in a group and link against `asset1`:
```
group2 = AssetGroup([asset2]).prefixed("some_other_prefix", upstream_groups=[group1])
```

`prefixed` will notice that `AssetKey(["prefix1", "asset1"])` is a suffix of `AssetKey(['prefix_2', 'prefix_1', 'asset1'])` and thus match them up. If `asset2` looked like this:

```
@asset(ins={"asset1": AssetIn(prefix="prefix2"))
def asset2(asset1):
    ...
```
then there would be no match, because `AssetKey(["prefix2", "asset1"])` is not a suffix of `AssetKey(['prefix_2', 'prefix_1', 'asset1'])` 

Does that answer your question? If not, is there a situation you have in mind where there would be more ambiguity?

**Comment by alangenfeld** [comment]
> I basically chose the name for consistency with configured

ah makes sense, i find that reasoning compelling 

>  I'm particularly un-enamored with assets_prefixed right now because it just feels kind of hard to parse, so this PR keeps it out of the public API

sounds like we are on the same page. Tricky to name the bare function variant. 

> Do you have thoughts?

The thinking was a path like `update_with(prefix_key=..., other_thing=...)` but i'm cool with a chaining end state.

File: python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_group.py (hunk 5 above)
Comment (sryza) [reply python_modules/dagster/dagster_tests/core_tests/asset_defs_tests/test_asset_group.py:581]: 👍

**Comment by sryza** [comment]
@OwenKephart @clairelin135 - inspired by some of your feedback, I pushed what I think is a considerable simplification of this. Main changes from the previous version:
* No `upstream_assets` or `upstream_groups` argument.
* No fancy suffix-matching. If an input asset key exactly matches the asset key of another asset in the group, then the prefix is applied to that input asset key. Otherwise, it's not.

**Comment by clairelin135** [comment]
@sryza I find the latest set of changes much easier to understand, it seems intuitive to me that an AssetKey is unique so only `AssetIn`s with exactly matching keys will be matched. 

I also like that `upstream_assets` is no longer an argument of `assets_prefixed`, since the same prefixing could be achieved  using just the `assets` argument.

**Comment by OwenKephart** [comment]
@sryza I like the new version of this, and after the PRs that I'm working on land, some remaining sketchy bits will go away. In particular, we won't have to mutate any of the OpDefinitions, as AssetKey information won't be stored there anymore.

I still think we might want to have some version of the fancy suffix matching in the future, but maybe this will only come into play once we actually try to create the job. Haven't thought through all the implications at all, and not super high priority.

**Comment by sryza** [comment]
@OwenKephart

> after the PRs that I'm working on land, some remaining sketchy bits will go away. In particular, we won't have to mutate any of the OpDefinitions, as AssetKey information won't be stored there anymore.

is there a branch that I should rebase this on top of?

**Comment by OwenKephart** [comment]
@sryza I think https://github.com/dagster-io/dagster/pull/7473 should work

**Comment by sryza** [comment]
@OwenKephart thanks - rebased

**Review by OwenKephart — APPROVED** [review]
likely best to rebase again on the top of the stack just to avoid conflicts: https://github.com/dagster-io/dagster/pull/7573. Hope to get those both merged in by early afternoon
